forc-plugins/forc-crypto/src/keys/vanity.rs

```diff
@@ -0,0 +1,551 @@
+    }
+}
+
+struct RightHexMatcher {
+    right: Vec<u8>,
+}
+
+impl VanityMatcher for RightHexMatcher {
+    fn is_match(&self, addr: &Address) -> bool {
+        let bytes = addr.as_ref();
+        bytes[bytes.len() - self.right.len()..]
+            .iter()
+            .zip(self.right.iter())
+            .all(|(a, b)| a.eq_ignore_ascii_case(b))
+    }
+}
+
+struct LeftExactRightRegexMatcher {
+    left: Vec<u8>,
+    right: Regex,
+}
+
+impl VanityMatcher for LeftExactRightRegexMatcher {
+    fn is_match(&self, addr: &Address) -> bool {
+        let bytes = addr.as_ref();
+        bytes.starts_with(&self.left) && self.right.is_match(&hex::encode(bytes))
+    }
+}
+
+struct LeftRegexRightExactMatcher {
+    left: Regex,
+    right: Vec<u8>,
+}
+
+impl VanityMatcher for LeftRegexRightExactMatcher {
+    fn is_match(&self, addr: &Address) -> bool {
+        let bytes = addr.as_ref();
+        self.left.is_match(&hex::encode(bytes))
+            && bytes[bytes.len() - self.right.len()..]
+                .iter()
+                .zip(self.right.iter())
+                .all(|(a, b)| a.eq_ignore_ascii_case(b))
+    }
+}
+
+struct SingleRegexMatcher {
+    re: Regex,
+}
+
+impl VanityMatcher for SingleRegexMatcher {
+    fn is_match(&self, addr: &Address) -> bool {
+        let addr = hex::encode(addr.as_ref());
+        self.re.is_match(&addr)
+    }
+}
+
+struct RegexMatcher {
+    left: Regex,
+    right: Regex,
+}
+
+impl VanityMatcher for RegexMatcher {
+    fn is_match(&self, addr: &Address) -> bool {
+        let addr = hex::encode(addr.as_ref());
+        self.left.is_match(&addr) && self.right.is_match(&addr)
+    }
+}
+
+fn parse_pattern(pattern: &str, is_start: bool) -> anyhow::Result<Either<Vec<u8>, Regex>> {
+    if pattern.len() > 64 {
+        return Err(VanityAddressError::InvalidPattern(
+            "Pattern too long: max 64 characters".to_string(),
+        )
+        .into());
+    }
+
+    if let Ok(decoded) = hex::decode(pattern) {
+        Ok(Either::Left(decoded))
+    } else {
+        // Check if the pattern contains only valid hex characters
+        if !pattern
+            .chars()
+            .all(|c| c.is_ascii_hexdigit() || c == '^' || c == '$')
+        {
+            return Err(VanityAddressError::InvalidPattern(format!("pattern: {}", pattern)).into());
+        }
+
+        // Case-insensitive regex pattern
+        let (prefix, suffix) = if is_start {
+            ("^(?i)", "")
+        } else {
+            ("(?i)", "$")
+        };
+        Ok(Either::Right(Regex::new(&format!(
+            "{prefix}{pattern}{suffix}"
+        ))?))
+    }
+}
+
+// Custom error type for vanity address generation
+#[derive(Debug)]
+pub enum VanityAddressError {
+    Timeout(u64),
+    GenerationFailed,
+    InvalidPattern(String),
+}
+
+impl fmt::Display for VanityAddressError {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            Self::Timeout(secs) => write!(
```

Comment:
We use the `tracing` library in forc and forc plugins to make it easier to silence output or provide verbose output in the CLIs. These `write!`s can be replaced with `tracing::error!`

Reply:
removed custom error as no longer needed as it did not see much usage; reverting back to anyhow: https://github.com/FuelLabs/sway/commit/80ca7a2ce4a87e0c8a6038ff69defc3e07af117d